File: app/src/main/java/com/tazzledazzle/stockagg/PriceProvider.java
package com.tazzledazzle.stockagg;

import java.util.concurrent.ThreadLocalRandom;
import java.util.function.Consumer;

/**
 * Simulates an upstream price provider. {@link #run} is a blocking loop
 * meant to be submitted to a virtual-thread executor -- the blocking
 * {@code Thread.sleep} is the whole point: virtual threads make "just
 * write blocking code" viable at thousands-of-tasks scale, since the JVM
 * parks the virtual thread (cheaply) instead of pinning an OS thread.
 */
public class PriceProvider {

    private final String name;
    private final double basePrice;
    private final Integer flakyEveryNTicks; // null = never stalls

    public PriceProvider(String name, double basePrice, Integer flakyEveryNTicks) {
        this.name = name;
        this.basePrice = basePrice;
        this.flakyEveryNTicks = flakyEveryNTicks;
    }

    public String name() {
        return name;
    }

    /** Blocking loop. Intended to run on its own virtual thread. */
    public void run(String symbol, Consumer<Tick> onTick) {
        double price = basePrice;
        int tickCount = 0;
        try {
            while (!Thread.currentThread().isInterrupted()) {
                tickCount++;

                boolean flakyTick = flakyEveryNTicks != null && tickCount % flakyEveryNTicks == 0;
                long delayMillis = flakyTick ? 5_000L : ThreadLocalRandom.current().nextLong(200, 800);
                Thread.sleep(delayMillis);

                price += ThreadLocalRandom.current().nextDouble(-0.5, 0.5);
                price = Math.max(price, 0.01);

                onTick.accept(new Tick(symbol, name, price, System.currentTimeMillis()));
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
}

File: app/src/main/kotlin/com/tazzledazzle/stockagg/PriceProvider.kt
package com.tazzledazzle.stockagg

import com.tazzledazzle.stockagg.model.Tick
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlin.random.Random

/**
 * Simulates an upstream price provider (in real life: a WebSocket feed,
 * FIX connection, or REST poller against Alpha Vantage / Polygon / etc).
 *
 * Modeled as a suspending [Flow] so it composes naturally with structured
 * concurrency: cancelling the collector cancels the underlying "connection"
 * (the delay loop) automatically -- no manual thread interruption needed.
 *
 * A provider can be "flaky" to demonstrate the bulkhead behavior: a flaky
 * provider occasionally stalls far longer than the timeout, and the
 * aggregator must not let that block other providers or other symbols.
 */
class PriceProvider(
    val name: String,
    private val basePrice: Double,
    private val flakyEveryNTicks: Int? = null
) {
    fun ticks(symbol: String): Flow<Tick> = flow {
        var price = basePrice
        var tickCount = 0
        while (true) {
            tickCount++

            // Simulate an occasional slow/stalled provider.
            val isFlakyTick = flakyEveryNTicks != null && tickCount % flakyEveryNTicks == 0
            val delayMillis = if (isFlakyTick) 5_000L else Random.nextLong(200, 800)
            delay(delayMillis)

            // Random walk the price.
            price += Random.nextDouble(-0.5, 0.5)
            price = price.coerceAtLeast(0.01)

            emit(
                Tick(
                    symbol = symbol,
                    provider = name,
                    price = price,
                    timestampMillis = System.currentTimeMillis()
                )
            )
        }
    }
}